Add Compression Dictionary Transport docs (#38722)
mdn/content: files/en-us/web/http/reference/headers/use-as-dictionary/index.md

---
title: Use-As-Dictionary
slug: Web/HTTP/Reference/Headers/Use-As-Dictionary
page-type: http-header
browser-compat: http.headers.Use-As-Dictionary
---

{{HTTPSidebar}}

The HTTP **`Use-As-Dictionary`** response header lists the matching criteria that the {{glossary("Compression Dictionary Transport")}} dictionary can be used for, for future requests.

See the [Compression Dictionary Transport guide](/en-US/docs/Web/HTTP/Guides/Compression_dictionary_transport) for more information.

## Syntax

```http
Use-As-Dictionary: match="<urlpattern>"
Use-As-Dictionary: match-dest=("<destination1>" "<destination2>", …)
Use-As-Dictionary: id="<string-identifier>"
Use-As-Dictionary: type="raw"

// Multiple, in any order
Content-Encoding: match="<urlpattern>", match-dest=("<destination1>")
```

## Directives

- `match`
  - : A string value containing a [URL Pattern](/en-US/docs/Web/API/URL_Pattern_API): only resources whose URLs match this pattern may use this resource as a dictionary.
- `match-dest`

>   - : A space-separated list of strings, with each string in quotres and the whole value enclosed in parentheses, that provides a list of [Fetch request destinations](/docs/Web/API/Request/destination) for the dictionary to match.

Reviewer: ```suggestion
  - : A space-separated list of strings, with each string in quotes and the whole value enclosed in parentheses, that provides a list of [Fetch request destinations](/docs/Web/API/Request/destination) that requests must match if they are to use this dictionary.
```

(I think?) The dictionary doesn't match, the request does.

---
title: Use-As-Dictionary
slug: Web/HTTP/Reference/Headers/Use-As-Dictionary
page-type: http-header
browser-compat: http.headers.Use-As-Dictionary
---

{{HTTPSidebar}}

The HTTP **`Use-As-Dictionary`** header lists the matching criteria that the {{glossary("Compression Dictionary Transport")}} dictionary can be used for, for future requests.

## Directives

- `match`

>   - : A string value that provides the {{domxref("URL Pattern API", "URL Pattern")}} to use for future request matching.

Reviewer 2: ```suggestion
  - : A string value that provides the {{domxref("URL Pattern API", "URL Pattern")}} to use for future request matching (regex patterns are not supported).
```

>   - : A string value that provides the {{domxref("URL Pattern API", "URL Pattern")}} to use for future request matching (regex patterns are not supported).

Reviewer: ```suggestion
  - : A string value containing a [URL Pattern](/en-US/docs/Web/API/URL_Pattern_API): only resources whose URLs match this pattern may use this resource as a dictionary.
```

("URL Pattern" is not an identifier so it should not get code formatting).

I find the URL Pattern docs really hard to understand tbh, but that's outside the scope of this PR.

- `match-dest`

>   - : An [Inner List of String values](https://www.rfc-editor.org/rfc/rfc8941#name-inner-lists) that provides a list of [Fetch request destinations](/docs/Web/API/Request/destination) for the dictionary to match.

Reviewer: I don't understand "Inner List of String values" and don't think we can give people a link that will make them understand ABNF. Is this actually effectively just a space-separated list of strings, with each string in quotres and the whole value enclosed in parentheses?

Author: Yup. Changed it to that.

- `id`

>   - : A string value that specifies a server identifier for the dictionary. This will then be added as a {{HTTPHeader("Dictionary-ID")}} to resource requests where that dictionary can be used for.

Reviewer: ```suggestion
  - : A string value that specifies a server identifier for the dictionary. This ID value will then be added in the  {{HTTPHeader("Dictionary-ID")}} request header when the browser requests a resource which can use this dictionary.
```

- `type`

>   - : A token value that describes the file format of the supplied dictionary. Currently only `raw` is supported (which is the default) so this is more for future compatibility.

Reviewer: ```suggestion
  - : A string value that describes the file format of the supplied dictionary. Currently only `raw` is supported (which is the default) so this is more for future compatibility.
```

?

## Examples

> [!NOTE]

> > See the [Compression Dictionary Transport guide](/en-US/docs/Web/HTTP/Guides/Compression_dictionary_transport) for more information.

Reviewer: Again put this in the general description, not with examples.

Author: Done.

### Path prefix

```http
Use-As-Dictionary: match="/product/*"
```

This says the dictionary is only to be used for URLs starting with `/product/`.

### Versioned directories

```http
Use-As-Dictionary: match="/app/*/main.js"
```

This uses a wildcard to match multiple versions of a file.

### Destinations

```http
Use-As-Dictionary: match="/product/*", match-dest=("document")
```

> This uses`match-dest` to ensure the dictionary is inly used for `document` requests so `/product/js/app.js` would not match and hence would not use this dictionary.

Reviewer: ```suggestion
This uses`match-dest` to ensure the dictionary is only used for `document` requests, so `/product/js/app.js` would not match and hence would not use this dictionary.
```

But also, isn't it more accurate to say "resource requests originating from a `<script>` tag"  - i.e. it's the way the resource is requests, not its extension, that determines `destination`?

Author: Yup! Changed it to:

> This uses`match-dest` to ensure the dictionary is only used for `document` requests so `<script src="/product/js/app.js">` resource requests for example would not match.

```http
Use-As-Dictionary: match="/product/*", match-dest=("document" "frame")
```

This would allow the dictionary to match both top-level documents and iframes.

> [!NOTE]

> > As a [structured field](https://www.rfc-editor.org/rfc/rfc8941), the parameters should be space-separated, quoted strings — as shown above — and not comma-separated, which developers may be more used to.

Reviewer: This should be in Directives, not here. Also I would just describe the syntax directly: "The value of `match-dest` should be a space-separated list of quoted strings, enclosed in parentheses".

Author: Removed.

### Id

```http
Use-As-Dictionary: match="/product/*", id="dictionary-12345"
```

> In which case the `id` will be used in the future requests in the {{HTTPHeader("Dictionary-ID")}} HTTP header along with the SHA-256 hash of the dictionary surrounded by colons in the {{HTTPHeader("Available-Dictionary")}} HTTP header:

Reviewer: ```suggestion
When `Use-As-Dictionary` includes an `id` directive, as in this example, the `id` value will be included in the {{HTTPHeader("Dictionary-ID")}} request header for resources that can use this dictionary. The resource request will also include the SHA-256 hash of the dictionary surrounded by colons in the {{HTTPHeader("Available-Dictionary")}} header:
```
